### PR TITLE
Implement a `parquet` parser and printer

### PR DESCRIPTION
**Pull request by balavinaithirthan**
We developed a plugin for Tenzir that parses and prints data in the parquet format, similar to other formats like JSON, feather, and YAML. Our parquet parser and printer enable parsing and printing through operators such as write, to, and from. Parquet requires a single schema. The parser is implemented in a blocking format and receives an upstream generator of chunk_ptr (byte representation) as input, which returns a generator of table_slice for downstream operators. Within the parser, the complete parquet input with metadata is collected and converted into record batches before being co_yielded as a table_slice.
In contrast, the printer is a streaming class that takes a generator of table slices as input and outputs a generator of chunks. Table slices are converted to record batches and written as parquet, with the parquet footer written at the end based on the metadata from the stream. To create valid metadata, we derived a new arrow::io::OutputStream that acts as a file for metadata purposes and a stream for memory efficiency. The generator of table_slice input is written to a byte buffer before being co_yielded at the end of a row.

Change log: The `parquet` format now reads and writes Arrow IPC streams and Parquet files. The format is streamable for `write parquet`.

### CODE REVIEW

File: libtenzir/src/contiguous_buffer_stream.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+#include <tenzir/contiguous_buffer_stream.hpp>
+
+#include <arrow/array.h>
+#include <arrow/compute/cast.h>
+#include <arrow/io/api.h>
+#include <arrow/io/interfaces.h>
+#include <arrow/table.h>
+
+#include <memory>
```
Comment (balavinaithirthan) [inline libtenzir/src/contiguous_buffer_stream.cpp:14]: Clean up #include statements

File: libtenzir/src/contiguous_buffer_stream.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  capacity_ = initial_capacity;
+  streaming_capacity_ = initial_capacity;
+  buffer_position_ = 0;
+  streaming_position_ = 0;
+  mutable_data_ = buffer_->mutable_data();
+  return arrow::Status::OK();
+}
+
+auto contiguous_buffer_stream::PurgeReset() -> arrow::Status {
+  int64_t initial_capacity = 1024;
```
Comment (balavinaithirthan) [inline libtenzir/src/contiguous_buffer_stream.cpp:64]: Still working through not reallocating every time

File: libtenzir/src/contiguous_buffer_stream.cpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  capacity_ = initial_capacity;
+  buffer_position_ = 0;
+  mutable_data_ = buffer_->mutable_data();
+  return arrow::Status::OK();
+}
+
+contiguous_buffer_stream::~contiguous_buffer_stream() {
+  if (buffer_) {
+    // arrow::io::internal::CloseFromDestructor(this);
+    arrow::Status st = this->Close();
```
Comment (balavinaithirthan) [inline libtenzir/src/contiguous_buffer_stream.cpp:77]: How do we error catch here without internal arrow library?

File: libtenzir/src/contiguous_buffer_stream.cpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    //     std::stringstream ss;
+    //     ss << "When destroying file of type " << file_type << ": " <<
+    //     st.message(); ARROW_LOG(FATAL) << st.WithMessage(ss.str());
+    // #endif
+  }
+}
+
+auto contiguous_buffer_stream::Purge()
+  -> arrow::Result<std::shared_ptr<arrow::Buffer>> {
+  auto sliced_buffer_result
```
Comment (balavinaithirthan) [inline libtenzir/src/contiguous_buffer_stream.cpp:94]: This returns a view of the original buffer as a slice.

File: libtenzir/src/contiguous_buffer_stream.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+  int64_t new_capacity = std::max(kBufferMinimumSize, capacity_);
+  while (new_capacity < buffer_position_ + nbytes) {
+    new_capacity = new_capacity * 2;
+  }
+  if (new_capacity > capacity_) {
+    RETURN_NOT_OK(buffer_->Resize(new_capacity));
+    capacity_ = new_capacity;
+    mutable_data_ = buffer_->mutable_data();
+  }
+
```
Comment (balavinaithirthan) [inline libtenzir/src/contiguous_buffer_stream.cpp:163]: Remove whitespace

File: libtenzir/src/contiguous_buffer_stream.cpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+auto contiguous_buffer_stream::Streaming_Reserve(int64_t nbytes)
+  -> arrow::Status {
+  int64_t new_capacity = std::max(kBufferMinimumSize, streaming_capacity_);
+  while (new_capacity < streaming_position_ + nbytes) {
+    new_capacity = new_capacity * 2;
+  }
+  if (new_capacity > streaming_capacity_) {
+    streaming_capacity_ = new_capacity;
+  }
+
```
Comment (balavinaithirthan) [inline libtenzir/src/contiguous_buffer_stream.cpp:176]: remove whitespace

File: plugins/parquet/parquet.cpp [hunk 7]
```diff
@@ -23,6 +29,10 @@
 #include <parquet/arrow/reader.h>
 #include <parquet/arrow/writer.h>
 
+#include <cstdint>
+#include <memory>
+#include <string>
+
```
Comment (balavinaithirthan) [inline plugins/parquet/parquet.cpp:35]: clean up include statements

File: plugins/parquet/parquet.cpp [hunk 8]
```diff
@@ -475,10 +507,173 @@ class active_parquet_store final : public active_store {
...
+    return false;
+  }
+
+  friend auto inspect(auto& f, parquet_printer& x) -> bool {
+    return f.object(x).fields();
+  }
+
+  class parquet_printer_instance : public printer_instance {
+  public:
+    static auto make(operator_control_plane& ctrl, type input_schema)
```
Comment (balavinaithirthan) [inline plugins/parquet/parquet.cpp:568]: How to error handle in this make function?

**Review by balavinaithirthan — COMMENTED** [review]
Initial passthrough

File: plugins/parquet/parquet.cpp [hunk 9]
```diff
@@ -475,10 +507,173 @@ class active_parquet_store final : public active_store {
...
+    auto parser = argument_parser{"feather", "https://docs.tenzir.com/next/"
+                                             "formats/feather"};
+    parser.parse(p);
+    return std::make_unique<parquet_parser>();
+  }
+
+  auto parse_printer(parser_interface& p) const
+    -> std::unique_ptr<plugin_printer> override {
+    auto parser = argument_parser{"feather", "https://docs.tenzir.com/next/"
+                                             "formats/feather"};
```
Comment (dominiklohmann) [inline plugins/parquet/parquet.cpp:671]: s/feather/parquet

File: plugins/parquet/parquet.cpp [hunk 10]
```diff
@@ -475,10 +507,173 @@ class active_parquet_store final : public active_store {
...
+      return diagnostic::error("{}", out_buffer_result.status().ToString())
+        .to_error();
+    }
+    auto out_buffer = out_buffer_result.MoveValueUnsafe();
+    auto file_result = ::parquet::arrow::FileWriter::Open(
+      *schema, arrow::default_memory_pool(), out_buffer);
+    if (!file_result.ok()) {
+      return diagnostic::error("{}", file_result.status().ToString()).to_error();
+    }
+    auto writer = file_result.MoveValueUnsafe();
```
Comment (dominiklohmann) [inline plugins/parquet/parquet.cpp:554]: This can all go.

File: plugins/parquet/parquet.cpp [hunk 11]
```diff
@@ -475,10 +507,173 @@ class active_parquet_store final : public active_store {
...
-class plugin final : public virtual store_plugin {
-  caf::error initialize(const record& plugin_config,
-                        const record& global_config) override {
+class parquet_printer final : public plugin_printer {
+public:
+  auto name() const -> std::string override {
+    return "parquet";
+  }
+
+  auto instantiate([[maybe_unused]] type input_schema,
```
Comment (dominiklohmann) [inline plugins/parquet/parquet.cpp:539]:
```suggestion
  auto instantiate(type input_schema,
```

File: plugins/parquet/parquet.cpp (hunk 8 above)
Comment (dominiklohmann) [reply plugins/parquet/parquet.cpp:568]: Let's pair on this.

File: libtenzir/include/tenzir/contiguous_buffer_stream.hpp
Comment (dominiklohmann) [inline libtenzir/include/tenzir/contiguous_buffer_stream.hpp:1]: Let's move this into the Parquet plugin.

File: libtenzir/include/tenzir/contiguous_buffer_stream.hpp [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+#include "arrow/io/file.h"
+
+#include <arrow/array.h>
+#include <arrow/compute/cast.h>
+#include <arrow/io/api.h>
+#include <arrow/table.h>
+
+namespace tenzir {
+
+class contiguous_buffer_stream : public arrow::io::OutputStream {
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/contiguous_buffer_stream.hpp:12]: For maintainability, we should consider inheriting from BufferOutputStream instead.

File: .gitignore [hunk 13]
```diff
@@ -32,5 +32,6 @@ venv/
 # Python build artifacts
 /python/dist
 
-# vscode build artifacts
-.vscode/launch.json
\ No newline at end of file
+# VScode built artifacts
+/.vscode/
+out/build/Test Preset/.cmake/api/v1/query/client-vscode/
```
Comment (balavinaithirthan) [inline .gitignore:37]: .cmake/
cmake build artifacts

File: changelog/next/changes/4116--parquet.format.md
Comment (balavinaithirthan) [inline changelog/next/changes/4116--parquet.format.md:1]: change to parquet-format.md

File: plugins/parquet/integration/tests/tests.bats [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+  size2=$(filesize "$file2")
+  check [ "$size1" -gt "$size2" ]
+}
+
+@test "truncated input" {
+  file=$(mktemp)
+  check tenzir "from ${BATS_TENZIR_DATADIR}/inputs/zeek/conn.log.gz read zeek-tsv | batch 512 | to ${file} write parquet"
+  dd "if=${file}" "of=${file}.10k" bs=1 count=10000
+  dd "if=${file}" "of=${file}.1M" bs=1 count=1000000
+  check ! tenzir "from ${file}.10k read parquet | summarize count(.)"
```
Comment (balavinaithirthan) [inline plugins/parquet/integration/tests/tests.bats:53]: rewrite test for parquet specific truncation without magic bytes in footer

File: plugins/parquet/parquet.cpp [hunk 15]
```diff
@@ -6,511 +6,326 @@
...
-    slice.offset(detail::narrow_cast<id>(offset));
+  auto input_file = as_arrow_file(std::move(parquet_chunk));
+  auto parquet_reader_properties
+    = ::parquet::ReaderProperties(arrow::default_memory_pool());
+  parquet_reader_properties.enable_buffered_stream();
+  std::unique_ptr<::parquet::arrow::FileReader> out_buffer;
+  auto arrow_reader_properties = ::parquet::ArrowReaderProperties();
+
+  // 2^16 is the largest batch size allowed in a table slice
+  arrow_reader_properties.set_batch_size(65536);
```
Comment (balavinaithirthan) [inline plugins/parquet/parquet.cpp:83]:
```suggestion
  arrow_reader_properties.set_batch_size(defaults::import::table_slice_size);
```

(and remove the comment)

File: plugins/parquet/parquet.cpp [hunk 16]
```diff
@@ -6,511 +6,326 @@
...
+  if (!record_batch_reader_status.ok()) {
+    diagnostic::error("{}", record_batch_reader_status.ToString())
+      .note("failed create record batches from input data")
+      .emit(ctrl.diagnostics());
+    co_return;
+  }
+  for (arrow::Result<std::shared_ptr<arrow::RecordBatch>> maybe_batch :
+       *rb_reader) {
+    if (!maybe_batch.ok()) {
+      diagnostic::error("{}", maybe_batch.status().ToString())
```
Comment (balavinaithirthan) [inline plugins/parquet/parquet.cpp:114]: Needs to be ToStringWithoutContextLines here and above, whenever we're creating a diagnostic

**Review by dominiklohmann — APPROVED** [review]
We reviewed this together. This is a really great piece of work, and I'm looking forward to having it soon.

**Pull request merged**